operator/pkg/ciliumendpointslice/manager.go

```diff
@@ -0,0 +1,635 @@
+			}
+		}
+	}
+}
+
+// If available, getCESFromCache returns CiliumEndpointSlice object.
+func (c *cesMgr) getCESFromCache(cesName string) (*cilium_v2.CiliumEndpointSlice, error) {
+	if ces, exists := c.desiredCESs.getCESTracker(cesName); exists {
+		return ces.ces, nil
+	}
+	return nil, fmt.Errorf("Failed to get CES from local cache for the CESName: %s", cesName)
+}
+
+// getCESCopyFromCache returns the copy of CiliumEndpointSlice object.
+func (c *cesMgr) getCESCopyFromCache(cesName string) (*cilium_v2.CiliumEndpointSlice, error) {
+	if ces, exists := c.desiredCESs.getCESTracker(cesName); exists {
+		outCES := new(cilium_v2.CiliumEndpointSlice)
+		ces.backendMutex.RLock()
+		ces.ces.DeepCopyInto(outCES)
+		ces.backendMutex.RUnlock()
+		return outCES, nil
+
+	}
+	return nil, fmt.Errorf("Failed to get CES Copy from local cache for the CESName: %s", cesName)
+}
+
+// InsertCEPInCache is used to insert CEP in local cache, this may result in creating a new
+// CES object or updating an existing CES object.
+func (c *cesMgr) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string) string {
+	log.WithFields(logrus.Fields{
+		"CEPName": GetCEPNameFromCCEP(cep, ns),
+	}).Debug("Insert CEP in local cache")
+
+	// check the given cep is already exists in any of the CES.
+	// if yes, Update a ces with the given cep object.
+	if cesName, exists := c.desiredCESs.getCESName(GetCEPNameFromCCEP(cep, ns)); exists {
+		// add a cep into the ces
+		c.addCEPtoCES(cep, cesName)
+		return cesName
+	}
+
+	// If given cep object isn't packed in any of the CES. find a new ces
+	// to pack this cep.
+	cb := func() *cesTracker {
+		// get first available CES
+		for _, ces := range c.desiredCESs.getAllCESs() {
+			ces.backendMutex.RLock()
+			if ces.ces.Namespace != ns || len(ces.ces.Endpoints) >= c.maxCEPsInCES || len(ces.ces.Endpoints) == 0 {
+				ces.backendMutex.RUnlock()
+				continue
+			}
+			ces.backendMutex.RUnlock()
+			return ces
+		}
+		// allocate a new cesTracker and return
+		newCES := c.createCES("")
+		// Update the namespace to CES
+		newCES.ces.Namespace = ns
+		return newCES
+	}()
+
+	// Cache CEP name with newly allocated CES.
+	c.desiredCESs.insertCEP(GetCEPNameFromCCEP(cep, ns), cb.ces.GetName())
+
+	// Queue the CEP in CES
+	c.addCEPtoCES(cep, cb.ces.GetName())
+	return cb.ces.GetName()
+}
+
+// RemoveCEPFromCache is used to remove the CEP from local cache, this may result in
+// Updating an existing CES object.
+func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
+	log.WithFields(logrus.Fields{
+		"CEPName": cepName,
+	}).Debug("Remove CEP from local cache")
+
+	// Check in local cache, if a given cep is already batched in one of the ces.
+	// and if exists, delete cep from ces.
+	if cesName, exists := c.desiredCESs.getCESName(cepName); exists {
+		var ces *cesTracker
+		if ces, exists = c.desiredCESs.getCESTracker(cesName); !exists {
+			log.Info("Valid CEP-CES mapping, but CES doesn't exist in CES cache.")
```

Comment:
Is this log msg enough context for it to be informational to the user? I think it would help to say

```suggestion
			log.Info("Attempted to remove non-existent CES, skipping.")
```

The idea is that each time we log something INFO level or greater, it's a signal to the user. So log msgs should be actionable or at least inform the user what Cilium has decided to do & why or what the user should do to rectify the situation.

Reply:
I don't know how much user can understand from this message, but this message is for developers. We have couple of mapping. One CEP-to CES mapping and other one inverse of previous mapping i.e. CES-to-CEPs. This message says it found a hit on CEP-to-CES mapping but not on other map.

Reply:
If the message is for developers, is that to signal that there is a bug and developers should introduce a change to mitigate this? If so, then saying something like "Please report this bug to the developers" seems reasonable.

Reply:
AFAIK, I really don't think there is any bug here. During early development phase, we have seen some cases like that, but those issues are resolved and fixed. It was added to out of an abundance caution. I'm updating the message as per earlier suggestion. 

Adding `Please report this bug to the developers` comments looks okay to me, but i makes a case where we have a bug in code. In reality we don't.